### PR TITLE
Update minimum compiler versions + clean up

### LINKED ISSUE #5285: Minimum Compiler Version Update
After the 3.7 release, Kokkos is requiring C++17, which necessitates an upgrade in minimum compiler versions.
Furthermore, we are planning to move HIP out of Experimental in the release after 3.7 and start more stable support for ROCM versions. The question arises what that means for minimum versions Kokkos should support.

Here is the proposal we are working with right now, however we are considering lifting the version a bit higher than what we could get away with. The main reason to do so is avoiding known bugs, as well as making our burden for testing/support a bit easier.

| Compiler Line | Current Minimum | Initial Proposal | Considering | Decision | 
| --- | --- | --- | --- | --- | 
| GCC | 5.3 | 7.3 | **8.2** | 8.2 | 
| Clang | 4.0 | **8.0** | 11.1 | 8.0 | 
| Clang as CUDA compiler | 10.0 | **10.0** | 11.1 | 10.0 | 
| Intel `icpc` | 17.0 | 19.1 | **2019.5** (aka `icpc --version` 19.0.5) | 2019.5 | 
| CUDA `nvcc` | 9.2.88 | **11.0** | 11.x | 11.0 |
| Minimum CUDA for use with Clang as CUDA compiler | 10.0.1 | 10.0.1 | -- | 10.0.1 |
| ROCM | 4.5.0 | 5.2.0 | -- | 5.2.0 | 
| IntelLLVM (CPU) `icpx` | 2021.1.1 | 2021.1.1 | -- | 2021.1.1 |
| IntelLLVM (SYCL)| 2022.0.0 | 2022.2.0 | -- |  2022.2.0 |
| NVHPC `nvc++` | 22.3 | 22.3 | -- | 22.3 |
| MSVC | | | | 19.29 |
| IBM XL |  | | | Not supported |
| PGI | | | | now via NVC++ |

Couple notes: 
- Clang as CUDA compiler requires at least 10.x possibly 11.x we test 11.0 currently.
- NVC++ as CUDA compiler is not officially supported yet, we expect official support later this year
- The OpenMPTarget backend has separate requirements for newer compilers
- The upcoming OpenACC backend may have separate requirements for newer compilers

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #5285. This excludes ROCm/HIPCC and icpx. We'll update them later.

### CODE REVIEW

File: core/src/Kokkos_Macros.hpp [hunk 1]
```diff
@@ -213,12 +204,9 @@
...
 #define KOKKOS_ENABLE_PRAGMA_VECTOR 1
 #endif
-#if (1800 > KOKKOS_COMPILER_INTEL)
-#define KOKKOS_ENABLE_PRAGMA_SIMD 1
-#endif
 
-// FIXME Workaround for ICE with intel 17,18,19 in Trilinos
-#if (KOKKOS_COMPILER_INTEL <= 1900)
+// FIXME Workaround for ICE with intel <=21 in Trilinos
+#if (KOKKOS_COMPILER_INTEL <= 2100)
```
Comment (masterleinad) [inline core/src/Kokkos_Macros.hpp:209]: This corresponds to https://github.com/kokkos/kokkos/pull/5311.

File: core/src/Kokkos_Macros.hpp [hunk 2]
```diff
@@ -245,8 +233,8 @@
 #endif
 #endif
 
-#if (1700 > KOKKOS_COMPILER_INTEL)
-#error "Compiling with Intel version earlier than 17.0 is not supported."
+#if (1900 > KOKKOS_COMPILER_INTEL)
+#error "Compiling with Intel version earlier than 19.5 is not supported."
 #endif
```
Comment (masterleinad) [inline core/src/Kokkos_Macros.hpp:238]: Not sure about this. Can someone with access to 19.5 (aka 19.0.5) confirm the correct value here?

File: cmake/kokkos_compiler_id.cmake [hunk 3]
```diff
@@ -147,33 +147,33 @@ ENDIF()
 
 # Enforce the minimum compilers supported by Kokkos.
 SET(KOKKOS_MESSAGE_TEXT "Compiler not supported by Kokkos.  Required compiler versions:")
-SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    Clang        4.0.0 or higher")
-SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    GCC          5.3.0 or higher")
-SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    Intel       17.0.0 or higher")
+SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    Clang        8.0.0 or higher")
+SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    GCC          8.2.0 or higher")
+SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    Intel       19.0.5 or higher")
```
Comment (masterleinad) [inline cmake/kokkos_compiler_id.cmake:152]: Not sure about this. Can someone with access to 19.5 (aka 19.0.5) confirm the correct value here?

File: cmake/kokkos_compiler_id.cmake [hunk 4]
```diff
@@ -147,33 +147,33 @@ ENDIF()
...
     MESSAGE(FATAL_ERROR "${KOKKOS_MESSAGE_TEXT}")
   ENDIF()
 ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL GNU)
-  IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 5.3.0)
+  IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 8.2.0)
     MESSAGE(FATAL_ERROR "${KOKKOS_MESSAGE_TEXT}")
   ENDIF()
 ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL Intel)
-  IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 17.0.0)
+  IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 19.0.5)
```
Comment (masterleinad) [inline cmake/kokkos_compiler_id.cmake:168]: Not sure about this. Can someone with access to 19.5 (aka 19.0.5) confirm the correct value here?

**Comment by PhilMiller** [comment]
I'll check the Intel version number stuff

**Comment by PhilMiller** [comment]
Intel 19.0.5:

```
[pbmille@kokkos-dev-2 ~]$ module load sems-intel/19.0.5

$ which icpc
/projects/sems/install/rhel7-x86_64/sems-compilers/tpl/intel/19.0.5/gcc/8.3.0/base/u4jh4al/compilers_and_libraries_2019.5.281/linux/bin/intel64/icpc

$ icpc --version
icpc (ICC) 19.0.5.281 20190815
Copyright (C) 1985-2019 Intel Corporation.  All rights reserved.

$ cmake -DCMAKE_CXX_COMPILER=icpc ~/repos/kokkos 
-- Setting default Kokkos CXX standard to 14
-- The CXX compiler identification is Intel 19.0.5.20190815
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /projects/sems/install/rhel7-x86_64/sems-compilers/tpl/intel/19.0.5/gcc/8.3.0/base/u4jh4al/compilers_and_libraries_2019.5.281/linux/bin/intel64/icpc - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- The project name is: Kokkos
-- SERIAL backend is being turned on to ensure there is at least one Host space. To change this, you must enable another host execution space and configure with -DKokkos_ENABLE_SERIAL=OFF or change CMakeCache.txt
-- Using -std=gnu++14 for C++14 extensions as feature
-- Built-in Execution Spaces:
--     Device Parallel: NoTypeDefined
--     Host Parallel: NoTypeDefined
--       Host Serial: SERIAL
-- 
-- Architectures:
-- Found TPLLIBDL: /usr/include  
-- Using internal desul_atomics copy
-- Kokkos Devices: SERIAL, Kokkos Backends: SERIAL
-- Configuring done
-- Generating done
-- Build files have been written to: /ascldap/users/pbmille/build/2022-08-11-icpc19

$ cmake -DCMAKE_CXX_STANDARD=17 -DCMAKE_CXX_COMPILER=icpc ~/repos/kokkos 
-- Setting default Kokkos CXX standard to 17
-- The CXX compiler identification is Intel 19.0.5.20190815
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /projects/sems/install/rhel7-x86_64/sems-compilers/tpl/intel/19.0.5/gcc/8.3.0/base/u4jh4al/compilers_and_libraries_2019.5.281/linux/bin/intel64/icpc - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- The project name is: Kokkos
-- SERIAL backend is being turned on to ensure there is at least one Host space. To change this, you must enable another host execution space and configure with -DKokkos_ENABLE_SERIAL=OFF or change CMakeCache.txt
-- Using -std=gnu++17 for C++17 extensions as feature
-- Built-in Execution Spaces:
--     Device Parallel: NoTypeDefined
--     Host Parallel: NoTypeDefined
--       Host Serial: SERIAL
-- 
-- Architectures:
-- Found TPLLIBDL: /usr/include  
-- Using internal desul_atomics copy
-- Kokkos Devices: SERIAL, Kokkos Backends: SERIAL
-- Configuring done
-- Generating done
-- Build files have been written to: /ascldap/users/pbmille/build/2022-08-11-icpc19

$ cat intel.cc 
#include <iostream>

int main() {
  std::cout << '"' << __INTEL_COMPILER << '"' << std::endl;
  return 0;
}

$ make CXX=icpc intel
icpc     intel.cc   -o intel

$ ./intel 
"1900"
```

File: cmake/kokkos_corner_cases.cmake [hunk 5]
```diff
@@ -46,15 +46,3 @@ IF(KOKKOS_CXX_COMPILER_ID STREQUAL AppleClang AND KOKKOS_ENABLE_OPENMP)
...
-IF (KOKKOS_CXX_COMPILER_ID STREQUAL GNU AND KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 7)
-  MESSAGE(FATAL_ERROR "The GCC version used is ${KOKKOS_CXX_COMPILER_VERSION} but GCC < 7 does not properly support *this capture. Please upgrade the compiler.")
-ENDIF()
-
-IF (KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA AND KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 11)
-  MESSAGE(FATAL_ERROR "Kokkos requires C++17 which NVCC only supports from version 11 on but the NVCC version used is ${KOKKOS_CXX_COMPILER_VERSION}.")
-ENDIF()
-IF (KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA AND KOKKOS_ENABLE_CUDA_CONSTEXPR AND KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 11.2)
-  MESSAGE(WARNING "You have requested -DKokkos_ENABLE_CUDA_CONSTEXPR=ON for NVCC ${KOKKOS_CXX_COMPILER_VERSION} which is known to trigger compiler bugs before NVCC version 11.2. See https://github.com/kokkos/kokkos/issues/3496")
-ENDIF()
```
Comment (PhilMiller) [inline cmake/kokkos_corner_cases.cmake:59]: We haven't set nvcc 11.2 as a requirement, so I think this needs to stay

File: core/src/impl/Kokkos_QuadPrecisionMath.hpp [hunk 6]
```diff
@@ -72,22 +72,16 @@ namespace Experimental {
 
 // clang-format off
 // Numeric distinguished value traits
-// Workaround GCC bug https://godbolt.org/z/qWb5oe4dx
-// error: '__builtin_huge_valq()' is not a constant expression
-#if defined(KOKKOS_COMPILER_GNU) && (KOKKOS_COMPILER_GNU >= 710)
```
Comment (PhilMiller) [inline core/src/impl/Kokkos_QuadPrecisionMath.hpp:77]: We were previously only defining these guarded traits for `defined(KOKKOS_COMPILER_GNU)` - I take it that was a pre-existing mistake, and should have been `!defined(GNU) || GNU>=710`?

File: core/unit_test/TestStringManipulation.cpp [hunk 7]
```diff
@@ -176,7 +170,6 @@ KOKKOS_FUNCTION constexpr bool test_strncat() {
 }
 static_assert(test_strncat());
 
-#if !defined(KOKKOS_COMPILER_GNU) && (KOKKOS_COMPILER_GNU >= 540)
```
Comment (PhilMiller) [inline core/unit_test/TestStringManipulation.cpp:179]: This condition was unsatisfiable - again, probably should have been `||`

File: cmake/kokkos_corner_cases.cmake (hunk 5 above)
Comment (masterleinad) [reply cmake/kokkos_corner_cases.cmake:59]: yes

File: core/src/impl/Kokkos_QuadPrecisionMath.hpp (hunk 6 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_QuadPrecisionMath.hpp:77]: That's what I think.

File: core/unit_test/TestStringManipulation.cpp (hunk 7 above)
Comment (masterleinad) [reply core/unit_test/TestStringManipulation.cpp:179]: Yes, I agree.

**Comment by masterleinad** [comment]
> Intel 19.0.5:

Thanks!

**Comment by masterleinad** [comment]
The only failure is
```
: [ RUN      ] default_exec.overlap_mdrange_policy
24: 
24: Kokkos::Cuda::initialize WARNING: Cuda is allocating into UVMSpace by default
24:                                   without setting CUDA_MANAGED_FORCE_DEVICE_ALLOC=1 or
24:                                   setting CUDA_VISIBLE_DEVICES.
24:                                   This could on multi GPU systems lead to severe performance"
24:                                   penalties.
24: 
24: Kokkos::Cuda::initialize WARNING: Cuda is allocating into UVMSpace by default
24:                                   without setting CUDA_MANAGED_FORCE_DEVICE_ALLOC=1 or
24:                                   setting CUDA_VISIBLE_DEVICES.
24:                                   This could on multi GPU systems lead to severe performance"
24:                                   penalties.
24: /var/jenkins/workspace/Kokkos/core/perf_test/PerfTest_ExecSpacePartitioning.cpp:395: Failure
24: Expected: (time_end) > (1.5 * time_overlap), actual: 0.00477586 vs 0.00496419
24: [  FAILED  ] default_exec.overlap_mdrange_policy (451 ms)
```
in `CUDA-11.0-NVCC-C++17-RDC`.

File: cmake/kokkos_compiler_id.cmake [hunk 8]
```diff
@@ -147,49 +147,66 @@ ENDIF()
...
   # Treat PGI internally as NVHPC to simplify handling both compilers.
   # Before CMake 3.20 NVHPC was identified as PGI, nvc++ is
   # backward-compatible to pgc++.
   SET(KOKKOS_CXX_COMPILER_ID NVHPC CACHE STRING INTERNAL FORCE)
+ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL "MSVC")
+  IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 19.29)
+    MESSAGE(FATAL_ERROR "${KOKKOS_MESSAGE_TEXT}")
+  ENDIF()
+ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL XL OR KOKKOS_CXX_COMPILER_ID STREQUAL XLClang)
+  MESSAGE(FATAL_ERROR "${KOKKOS_MESSAGE_TEXT}")
```
Comment (dalg24) [inline cmake/kokkos_compiler_id.cmake:209]: The list does not explicitly mention IBM XL
Should we say it is not supported?
Also should we have a flag to turn the error into a warning and continue processing?

File: core/src/Kokkos_Macros.hpp [hunk 9]
```diff
@@ -245,8 +222,8 @@
...
 #endif
 
-#if (1700 > KOKKOS_COMPILER_INTEL)
-#error "Compiling with Intel version earlier than 17.0 is not supported."
+#if (1900 > KOKKOS_COMPILER_INTEL)
```
Comment (dalg24) [inline core/src/Kokkos_Macros.hpp:225]: I suppose that means there is no fine grain way to differentiate patch releases.  Is that right?

File: core/src/impl/Kokkos_QuadPrecisionMath.hpp (hunk 6 above)
Comment (dalg24) [reply core/src/impl/Kokkos_QuadPrecisionMath.hpp:77]: Yes

File: core/unit_test/TestStringManipulation.cpp (hunk 7 above)
Comment (dalg24) [reply core/unit_test/TestStringManipulation.cpp:179]: My bad here too

File: core/src/Kokkos_Macros.hpp (hunk 9 above)
Comment (PhilMiller) [reply core/src/Kokkos_Macros.hpp:225]: There seems not to be, in source code, when using the 19.x line at least.

**Comment by PhilMiller** [comment]
Other than Damien's remark about explicitly reporting that IBM XL is no longer supported, this all looks good to me.

We may want to audit the remaining `#if` compiler version checks throughout the code for similar mistakes of boolean logic, since they seemed to be quite frequent. We may even want to introduce helper macros to encapsulate the correct comparisons, to avoid those mistakes in the future.

File: cmake/kokkos_compiler_id.cmake (hunk 8 above)
Comment (masterleinad) [reply cmake/kokkos_compiler_id.cmake:210]: See 30fc3d617.

**Review by cz4rs — APPROVED** [review]
Looks good to me.

**Comment by masterleinad** [comment]
Dropping XL is now part of #5349.

File: Makefile.kokkos [hunk 10]
```diff
@@ -227,26 +225,21 @@ endif
 
 # Set compiler warnings flags.
 ifeq ($(KOKKOS_INTERNAL_ENABLE_COMPILER_WARNINGS), 1)
-  ifeq ($(KOKKOS_INTERNAL_COMPILER_PGI), 1)
-    # TODO check if PGI accepts GNU style warnings
-    KOKKOS_INTERNAL_COMPILER_WARNINGS =
+  ifeq ($(KOKKOS_INTERNAL_COMPILER_CLANG), 1)
+    KOKKOS_INTERNAL_COMPILER_WARNINGS = -Wall -Wunused-parameter -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized
   else
```
Comment (dalg24) [inline Makefile.kokkos:230]: please format this way
```
ifeq (COND1)
  ...
else ifeq (COND2)
  ...
else ifeq (COND3)
...
else
  ...
endif
```
to improve readability

**Review by dalg24 — COMMENTED** [review]
Looks good besides the format fix I asked for

**Review by dalg24 — APPROVED** [review]

**Pull request merged**